### PR TITLE
ArduSub Merge

### PR DESCRIPTION
**Pull request by jaxxzer**
A less-fancy, more up-to-date PR for merging ArduSub. This is a continuation of https://github.com/ArduPilot/ardupilot/pull/5655. I am creating this second pull request in order to avoid force pushing to bluerobotics master.

### CODE REVIEW

File: ArduSub/defines.h [hunk 1]
```diff
@@ -454,11 +455,18 @@ enum ThrowModeState {
...
 
 // GCS failsafe definitions (FS_GCS_ENABLE parameter)
-#define FS_GCS_DISABLED                     0
-#define FS_GCS_ENABLED_ALWAYS_RTL           1
-#define FS_GCS_ENABLED_CONTINUE_MISSION     2
-
-// EKF failsafe definitions (FS_EKF_ACTION parameter)
+#define FS_GCS_DISABLED		0 // Disabled
+#define FS_GCS_WARN_ONLY	1 // Only send warning to gcs (only useful with multiple gcs links)
+#define FS_GCS_DISARM		2 // Disarm
```
Comment (khancyr) [inline ArduSub/defines.h]: minor correction to align everything!

File: ArduSub/events.cpp [hunk 2]
```diff
@@ -74,65 +74,84 @@ void Sub::failsafe_battery_event(void)
...
+	}
+
+	uint32_t tnow = AP_HAL::millis();
+
+	// We have a leak
+	// Always send a warning every 5 seconds
+	if(tnow > failsafe.last_leak_warn_ms + 5000) {
 		failsafe.last_leak_warn_ms = tnow;
-		gcs_send_text_fmt(MAV_SEVERITY_WARNING, "Leak Detected");
+		gcs_send_text(MAV_SEVERITY_WARNING, "Leak Detected");
```
Comment (khancyr) [inline ArduSub/events.cpp]: it could be interesting to log or notify when the first leak was detect.

File: libraries/SITL/SIM_Submarine.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
```
Comment (khancyr) [inline libraries/SITL/SIM_Submarine.cpp]: not needed as a global file handle this

File: ArduSub/radio.cpp [hunk 4]
```diff
@@ -61,6 +61,15 @@ void Sub::init_rc_in()
     // set default dead zones
     default_dead_zones();
 
+    // initialize rc input to 1500 on control channels (rather than 0)
+    for(int i = 0; i < 7; i++) {
+    	if(i == 4) {
```
Comment (khancyr) [inline ArduSub/radio.cpp]: miss a space after if. Nice trick ! I will add this to rover too !! will prevent some current problem!

File: libraries/SITL/SIM_Submarine.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    frame->calculate_forces(*this, input, rot_accel, body_accel);
+}
+    
+/*
+  update the Submarine simulation by one time step
+ */
+void Submarine::update(const struct sitl_input &input)
+{
+    // get wind vector setup
+    update_wind(input);
```
Comment (khancyr) [inline libraries/SITL/SIM_Submarine.cpp]: love that ^^

File: libraries/SITL/SIM_Submarine.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+#include <stdio.h>
+
+using namespace SITL;
+
+Submarine::Submarine(const char *home_str, const char *frame_str) :
+    Aircraft(home_str, frame_str),
+    frame(NULL)
+{
+    frame = Frame::find_frame(frame_str);
```
Comment (khancyr) [inline libraries/SITL/SIM_Submarine.cpp]: maybe this could be improve later

File: libraries/AP_HAL_SITL/SITL_cmdline.cpp [hunk 7]
```diff
@@ -91,6 +92,7 @@ static const struct {
     { "balloon",            Balloon::create },
     { "plane",              Plane::create },
     { "calibration",        Calibration::create },
+	{ "vectored",           Submarine::create },
```
Comment (khancyr) [inline libraries/AP_HAL_SITL/SITL_cmdline.cpp]: miss a space

File: Tools/autotest/sim_vehicle.py [hunk 8]
```diff
@@ -348,6 +349,12 @@ def wait_unlimited():
         "waf_target": "bin/ardurover",
         "default_params_filename": "default_params/rover-skid.parm",
     },
+    # SUB
+    "vectored": {
```
Comment (khancyr) [inline Tools/autotest/sim_vehicle.py]: better to have sub or sub-vectored to better identification and allow other sub type !

File: Tools/autotest/sim_vehicle.py [hunk 9]
```diff
@@ -348,6 +349,12 @@ def wait_unlimited():
...
+        "model": "vectored",
+        "waf_target": "bin/ardusub-vectored",
+        "default_params_filename": "Sub.parm",
```
Comment (khancyr) [inline Tools/autotest/sim_vehicle.py]: param should be move in default_params directory

File: ArduSub/defines.h (hunk 1 above)
Comment (khancyr) [inline ArduSub/defines.h]: forget, has been handle later

File: ArduSub/events.cpp (hunk 2 above)
Comment (khancyr) [inline ArduSub/events.cpp]: forget, has been handle later

File: ArduSub/radio.cpp (hunk 4 above)
Comment (khancyr) [inline ArduSub/radio.cpp]: forget, has been handle later

File: libraries/AP_HAL_SITL/SITL_cmdline.cpp (hunk 7 above)
Comment (khancyr) [inline libraries/AP_HAL_SITL/SITL_cmdline.cpp]: forget, has been handle later

File: libraries/SITL/SIM_Submarine.cpp (hunk 6 above)
Comment (jaxxzer) [inline libraries/SITL/SIM_Submarine.cpp]: Yes, I will have to work on this some more.

File: libraries/SITL/SIM_Submarine.cpp (hunk 5 above)
Comment (magicrub) [inline libraries/SITL/SIM_Submarine.cpp]: ha!

**Comment by magicrub** [comment]
370 commits is getting a bit crazy...

**Comment by tridge** [comment]
@magicrub I think its fine to have a large number of commits for what is really merging two projects.

**Comment by tridge** [comment]
@jaxxzer I think this PR could benefit from this one: https://github.com/ArduPilot/ardupilot/pull/5689

**Comment by jaxxzer** [comment]
@tridge I am very interested in using that, thanks for putting it together.
@magicrub I am cutting it down a bit, and removing the messier bits of our developments in libraries.

File: libraries/SITL/SIM_Submarine.cpp (hunk 3 above)
Comment (jaxxzer) [inline libraries/SITL/SIM_Submarine.cpp]: Done.

File: Tools/autotest/sim_vehicle.py (hunk 8 above)
Comment (jaxxzer) [inline Tools/autotest/sim_vehicle.py]: Done, thanks.

**Comment by tridge** [comment]
@jaxxzer I'm curious why you didn't subclass AC_AttitudeControl_Multi for AC_AttitudeControl_Sub. It looks like there is a fair bit of the multicopter attitude controller duplicated? It is fine to do it this way if you prefer it, I just thought it would be a cleaner approach

**Comment by tridge** [comment]
@jaxxzer you need to mark functions that you are overriding in AC_Poscontrol as being virtual.

**Comment by tridge** [comment]
also, please use the "override" keyword when overriding functions in a subclass

**Comment by jaxxzer** [comment]
>I'm curious why you didn't subclass AC_AttitudeControl_Multi for AC_AttitudeControl_Sub. It looks like there is a fair bit of the multicopter attitude controller duplicated?

It is an exact duplicate right now. I plan on removing or at least changing angle limiting and throttle mixing, but am leaving it as is for now so I don't break the functionality of Sub.

File: libraries/AP_Baro/AP_Baro.h [hunk 10]
```diff
@@ -107,6 +113,12 @@ class AP_Baro
     // used by Replay
     void setHIL(uint8_t instance, float pressure, float temperature, float altitude, float climb_rate, uint32_t last_update_ms);
 
+    // Set the primary baro
+    void set_primary_baro(uint8_t primary) { _primary_baro.set_and_save(primary); };
+
+    // Set the type (Air or Water) of a particular instance
+    void set_type(uint8_t instance, uint8_t type) { sensors[instance].type = type; };
```
Comment (OXINARF) [inline libraries/AP_Baro/AP_Baro.h]: Type should be *baro_type_t* instead of *uint8_t*

File: libraries/AP_Baro/AP_Baro.h [hunk 11]
```diff
@@ -147,6 +159,7 @@ class AP_Baro
     uint8_t _primary;
 
     struct sensor {
+        uint8_t type;                   // 0 for air pressure (default), 1 for water pressure
```
Comment (OXINARF) [inline libraries/AP_Baro/AP_Baro.h]: Same

Comment (jaxxzer) [inline libraries/AP_Baro/AP_Baro.h]: Done, Done and Done! As always, I appreciate your attention to detail and quality feedback. Thank you.

File: ArduSub/Parameters.cpp [hunk 12]
```diff
@@ -0,0 +1,982 @@
...
+    // @DisplayName: Position (horizonal) controller P gain
+    // @Description: Loiter position controller P gain.  Converts the distance (in the latitude direction) to the target location into a desired speed which is then passed to the loiter latitude rate controller
+    // @Range: 0.500 2.000
+    // @User: Standard
+    GGROUP(p_pos_xy,                "POS_XY_", AC_P),
+
+#if TRANSECT_ENABLED == ENABLED
+    GGROUP(pid_crosstrack_control, "XTRACK_", AC_PID),
+
+    GGROUP(pid_heading_control, "HEAD_", AC_PID),
```
Comment (jaxxzer) [inline ArduSub/Parameters.cpp]: This was an experimental thing, I've taken it out just now.

**Comment by OXINARF** [comment]
@jaxxzer I see you are having issues with signbit on Linux platforms. This is isinf (and others) all over again. If you want to learn how to fix it properly take a look at:

- https://github.com/ArduPilot/ardupilot/blob/master/Tools/ardupilotwaf/cxx_checks.py#L30
- https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Common/missing/cmath

**Comment by jaxxzer** [comment]
@OXINARF
>@jaxxzer I see you are having issues with signbit on Linux platforms. This is isinf (and others) all over again. If you want to learn how to fix it properly take a look at:

Thank you for the information. After looking through it, I decided to just check for 0 first, then compare the float to zero.

**Comment by jaxxzer** [comment]
@OXINARF 
> add a get type method so that Sub doesn't have to rely on the water one always being the second instance

Good idea, done.

>    setting the primary as a default instead of forcing it (that also applies to setting the baro noise in EKF)

I'm not sure if this could work, because the user may boot with or without an external depth sensor connected, and the settings will need to be changed accordingly. We could set defaults, but we will still have to force it for the case opposite of the defaults that we set.

**Comment by tridge** [comment]
I think this PR is ready to be merged to master. Any last minute objections?

**Comment by jaxxzer** [comment]
@rmackay9 was requested to review the change to mavlink rangefinder. I haven't heard from him on that yet. I can take that out and put in a separate pr.

**Comment by tridge** [comment]
the rangefinder change is good. I'll vouch for it.

**Comment by OXINARF** [comment]
> I'm not sure if this could work, because the user may boot with or without an external depth sensor connected, and the settings will need to be changed accordingly. We could set defaults, but we will still have to force it for the case opposite of the defaults that we set.

You can still put defaults according to having a water depth sensor or not, but using parameter defaults instead of forcing allows the user to try a different value. For example, let's say the user had two water baros connected, right now you force it to the first you find - I think the user can still change the primary parameter after that, but he needs to do it at every boot. Same if you don't have a water pressure sensor, you force it to be the first instance. And baro EKF noise parameter is the same too.
A suggestion: send warnings via MAVLink messages when you these parameters deviate from what you expect, but allow the user to change it as they wish. As I said in my review, these are changes that can happen in the future.

I agree with @tridge that this is ready :tada:

**Comment by OXINARF** [comment]
@jaxxzer The commits with messages *Sub: Use AHRS_VEHICLE_SUBMARINE as AHRS vehicle class* and *	Sub: Use default StorageManager layout instead of copter layout* have their messages swapped - functionality is correct, but the message is for the other commit.

**Comment by jaxxzer** [comment]
@OXINARF Fixed, thanks.

**Pull request merged**